analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsShouldBeLoggedOrThrown.cs

```diff
@@ -0,0 +1,67 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Common.Walkers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ExceptionsShouldBeLoggedOrThrown : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S2139";
+    private const string MessageFormat = "Either log this exception and handle it, or rethrow it with some contextual information.";
+    private const string LoggingStatementMessage = "Logging statement.";
+    private const string ThrownExceptionMessage = "Thrown exception.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(cc =>
+            {
+                if (cc.Compilation.ReferencesAny(KnownAssembly.MicrosoftExtensionsLoggingAbstractions, KnownAssembly.Log4Net, KnownAssembly.NLog, KnownAssembly.CastleCore, KnownAssembly.Serilog))
+                {
+                     cc.RegisterNodeAction(c =>
+                         {
+                             var catchClauseSyntax = (CatchClauseSyntax)c.Node;
+                             var walker = new CatchLoggingInvocationWalker(c.SemanticModel);
+                             if (catchClauseSyntax.Declaration?.Identifier is { } exceptionIdentifier // there is an exception to log
+                                 && catchClauseSyntax.DescendantNodes().OfType<ThrowStatementSyntax>().FirstOrDefault() is not null // and a throw statement (any; preliminary check)
+                                 && walker.SafeVisit(catchClauseSyntax)
+                                 && walker.IsExceptionLogged
+                                 && walker.LoggingInvocationWithException is { } loggingInvocationWithException
```

Comment:
This nullability check seems a bit redundant.
Based on your walker, `IsExceptionLogged` and `LoggingInvocationWithException`'s nullability is quantum-entangled, so since you check for `IsExceptionLogged` truthiness, this condition always evaluates to true.

Reply:
I wonder why the coverage is 100, the false condition should not be covered. The main reason I did the check here was to get the variable for location :)

Reply:
Good point, I do not know why coverage is trolling us 🤔